pkg/server/commands/reverseexpand/reverse_expand.go

```diff
@@ -267,39 +258,29 @@ func (c *ReverseExpandQuery) execute(
 
 	for _, edge := range edges {
 		innerLoopEdge := edge
-		if currentEdge != nil && currentEdge.Condition == graph.RequiresFurtherEvalCondition {
-			// propagate the condition to upcoming reverse expansions
-			// TODO don't mutate the edge, keep track of the previous edge's condition and use it in trySendCandidate
-			innerLoopEdge.Condition = graph.RequiresFurtherEvalCondition
-		}
+		intersectionOrExclusionInPreviousEdges := intersectionOrExclusionInPreviousEdges || innerLoopEdge.TargetReferenceIsIntersectionOrExclusion
 		subg.Go(func() error {
-			r := &reverseExpandRequest{
-				storeID:          storeID,
+			r := &ReverseExpandRequest{
+				StoreID:          req.StoreID,
+				ObjectType:       req.ObjectType,
+				Relation:         req.Relation,
+				User:             req.User,
+				ContextualTuples: req.ContextualTuples,
 				edge:             innerLoopEdge,
-				targetObjectRef:  targetObjRef,
-				sourceUserRef:    req.User,
-				contextualTuples: req.ContextualTuples,
 			}
 
 			switch innerLoopEdge.Type {
-			case graph.DirectEdge:
-				return c.reverseExpandDirect(subgctx, r, resultChan, resolutionMetadata)
+			case graph.DirectEdge, graph.TupleToUsersetEdge:
+				return c.readTuplesAndExecute(subgctx, r, resultChan, intersectionOrExclusionInPreviousEdges, resolutionMetadata)
 			case graph.ComputedUsersetEdge:
-				// lookup the rewritten target relation on the computed_userset ingress
-				return c.execute(subgctx, &ReverseExpandRequest{
-					StoreID:    storeID,
-					ObjectType: req.ObjectType,
-					Relation:   req.Relation,
-					User: &UserRefObjectRelation{
-						ObjectRelation: &openfgav1.ObjectRelation{
-							Object:   sourceUserObj,
-							Relation: innerLoopEdge.TargetReference.GetRelation(),
-						},
+				// follow the computed_userset edge
+				r.User = &UserRefObjectRelation{
+					ObjectRelation: &openfgav1.ObjectRelation{
+						Object:   sourceUserObj,
+						Relation: innerLoopEdge.TargetReference.GetRelation(),
 					},
-					ContextualTuples: r.contextualTuples,
-				}, resultChan, innerLoopEdge, resolutionMetadata)
-			case graph.TupleToUsersetEdge:
-				return c.reverseExpandTupleToUserset(subgctx, r, resultChan, resolutionMetadata)
+				}
+				return c.execute(subgctx, r, resultChan, intersectionOrExclusionInPreviousEdges, resolutionMetadata)
```

Comment:
With this change it should be straightforward to do this https://github.com/openfga/openfga/issues/974